File: src/main/kotlin/tracking/steps/Preprocessor.kt
package tracking.steps

import tracking.matchers.Factory

private const val CREATOR_NAME = "CREATOR_NAME"

class Preprocessor {
    private val cleaners = Factory.getCleaners()
    private val falsePositives = Factory.getFalsePositives()

    fun preprocess(input: String, creatorAliases: List<String>): String
    {
        var result = input

//        $contents = $this->applyFilters($url, $inputText); // TODO
//        $contents = $this->extractFromJson($contents); // TODO
        result = result.lowercase()
        result = cleaners.replaceIn(result)
        result = replaceCreatorAliases(result, creatorAliases)
        result = falsePositives.replaceIn(result)

        return result
    }

    private fun replaceCreatorAliases(input: String, aliases: List<String>): String {
        var result = input

        aliases.map(String::lowercase).forEach { alias ->
            result = result.replace(alias, CREATOR_NAME)

            if (alias.length > 2 && alias.endsWith("s")) {
                /* Thank you, English language, I am enjoying this */
                result = result.replace(alias.dropLast(1) + "'s", CREATOR_NAME, true)
            }
        }

        return result
    }
}

File: symfony/src/Tracking/Preprocessor.php
<?php

declare(strict_types=1);

namespace App\Tracking;

use App\Tracking\Patterns\Patterns;
use Veelkoov\Debris\StringList;

class Preprocessor
{
    private const int MAX_ANALYSED_SIZE_CHARACTERS = 1 * 1024 * 1024; // ~= 1 MiB (+multibyte characters)

    public function __construct(
        private readonly Patterns $patterns,
    ) {
    }

    public function preprocess(string $input, StringList $aliases): string
    {
        $result = mb_substr($input, 0, self::MAX_ANALYSED_SIZE_CHARACTERS);
        // TODO: URL strategy
        $result = strtolower($result);
        $result = $this->patterns->cleaners->do($result);
        $result = $this->replaceCreatorAliases($result, $aliases);
        $result = $this->patterns->falsePositives->do($result);

        return $result;
    }

    private function replaceCreatorAliases(string $input, StringList $aliases): string
    {
        $result = $input;

        foreach ($aliases as $alias) {
            $alias = strtolower($alias);

            $result = str_replace($alias, 'CREATOR_NAME', $result);

            if (mb_strlen($alias) > 2 && str_ends_with($alias, 's')) {
                $result = str_replace(substr($alias, 0, -1)."'s", 'CREATOR_NAME', $result);
            }
        }

        return $result;
    }
}
